homeassistant/helpers/llm.py

```diff
@@ -0,0 +1,234 @@
+                    hint = vol.Maybe(hints[1])
+                elif len(hints) == 2 and hints[1] is NoneType:
+                    hint = vol.Maybe(hints[0])
+                else:
+                    hint = vol.Any(
+                        *tuple(x if x is not NoneType else None for x in hints)
+                    )
+
+            schema[
+                vol.Required(param.name)
+                if param.default is inspect.Parameter.empty
+                else vol.Optional(param.name, default=param.default)
+            ] = hint
+
+        self.parameters = vol.Schema(schema)
+
+    async def async_call(self, hass: HomeAssistant, tool_input: ToolInput) -> Any:
+        """Call the function."""
+        kwargs = tool_input.tool_args
+        for parameter in inspect.signature(self.function).parameters.values():
+            if parameter.name == "hass":
+                kwargs["hass"] = hass
+            elif hasattr(ToolInput, parameter.name):
+                kwargs[parameter.name] = getattr(tool_input, parameter.name)
+
+        if inspect.iscoroutinefunction(self.function):
+            return await self.function(**kwargs)
+        return self.function(**kwargs)
+
+
+@callback
+def async_register_tool(hass: HomeAssistant, tool: Tool | Callable) -> None:
+    """Register an LLM tool with Home Assistant."""
+    if (tools := hass.data.get(DATA_KEY)) is None:
+        tools = hass.data[DATA_KEY] = {}
+
+    if not isinstance(tool, Tool):
+        tool = FunctionTool(tool)
+
+    if tool.name in tools:
+        raise HomeAssistantError(f"Tool {tool.name} is already registered")
+
+    tools[tool.name] = tool
+
+
+@callback
+def async_remove_tool(hass: HomeAssistant, tool: Tool | Callable | str) -> None:
+    """Remove an LLM tool from Home Assistant."""
+    if (tools := hass.data.get(DATA_KEY)) is None:
+        return
+
+    if isinstance(tool, str):
+        tool_name = tool
+    elif isinstance(tool, Tool):
+        tool_name = tool.name
+    else:
+        tool_name = tool.__name__
+        if tool_name.startswith("async_"):
+            tool_name = tool_name[len("async_") :]
+
+    tools.pop(tool_name, None)
+
+
+@callback
+def async_get_tools(hass: HomeAssistant) -> Iterable[Tool]:
+    """Return a list of registered LLM tools."""
+    tools: dict[str, Tool] = hass.data.get(DATA_KEY, {})
+
+    return tools.values()
+
+
+@callback
+async def async_call_tool(hass: HomeAssistant, tool_input: ToolInput) -> Any:
+    """Call a LLM tool, validate args and return the response."""
+    try:
+        tool = hass.data[DATA_KEY][tool_input.tool_name]
+    except KeyError as err:
+        raise HomeAssistantError(f'Tool "{tool_input.tool_name}" not found') from err
+
+    if tool_input.context is None:
+        tool_input.context = Context()
+
+    tool_input.tool_args = tool.parameters(tool_input.tool_args)
+
+    return await tool.async_call(hass, tool_input)
+
+
+def llm_tool(hass: HomeAssistant) -> Callable:
```

Comment:
Is there prior art in Home Assistant for decorators that accept `hass`? This seems like it will only work with inline functions which to me means that a decorator won't actually really work out in the end. (hass isn't around when declaring functions)

Reply:
Yes, it will work for inline functions. `async_added_to_hass` is a good place. If you define an inline function with this decorator, it will be executed at the time when the inline function is defined, which is within the call of the outer function.

Here are a couple of examples:
1. The `log_message` decorator in `mqtt` component: https://github.com/home-assistant/core/blob/f2460a697509d6171e8dd33cea68d57dce02ed0f/homeassistant/components/mqtt/humidifier.py#L303 https://github.com/home-assistant/core/blob/f2460a697509d6171e8dd33cea68d57dce02ed0f/homeassistant/components/mqtt/debug_info.py#L25
2. The `verify_domain_control` decorator: https://github.com/home-assistant/core/blob/f2460a697509d6171e8dd33cea68d57dce02ed0f/homeassistant/components/sonos/media_player.py#L116 https://github.com/home-assistant/core/blob/f2460a697509d6171e8dd33cea68d57dce02ed0f/homeassistant/helpers/service.py#L1106

Reply:
ok, i saw that as pretty limiting. Where are you planning to use this?

Reply:
I am not sure I plan to use it personally. Basically it is syntax sugar to provide the easiest way to define an LLM tool. I expect other integrations to use it.

Reply:
I think it is a good idea and we discussed it in discord i know, though perhaps we add back when we know we have a use for it.  I recommend we focus on getting the intent use cases working with the existing LLM integrations we have and add these later as we know of use cases.

Reply:
Yeah, we should remove all function calling from this initial implementation. Keep it as simple as possible and just wrap intents.

Reply:
Done